File: ports/csharp/src/core/kernel/BcsEvent.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;

namespace BCS.Core.Kernel {
    public class BcsEvent : IComparable<BcsEvent> {
        public int Type { get; set; }
        public int Priority { get; set; }

        public BcsEvent(int type, int priority = 0) {
            Type = type;
            Priority = priority;
        }

        public int CompareTo(BcsEvent other) {
            // Higher priority first
            return other.Priority.CompareTo(this.Priority);
        }
    }

    public class BcsObject {
        public BcsObject Parent { get; set; }
        public List<BcsObject> Children { get; set; }
        private List<Func<BcsEvent, bool>> _eventFilters;

        public BcsObject(BcsObject parent = null) {
            Parent = parent;
            Children = new List<BcsObject>();
            _eventFilters = new List<Func<BcsEvent, bool>>();
            if (parent != null) {
                parent.Children.Add(this);
            }
        }

        public void InstallEventFilter(Func<BcsEvent, bool> filter) {
            _eventFilters.Add(filter);
        }

        public virtual bool Event(BcsEvent e) {
            foreach (var filter in _eventFilters) {
                if (filter(e)) {
                    return true;
                }
            }
            return false;
        }
    }

    public class BcsEventLoop {
        // In C# .NET 10+ we can use PriorityQueue, but for broader compatibility a BlockingCollection with a concurrent wrapper or custom wait is common.
        // We'll use a BlockingCollection of a concurrent priority structure, or a simple Monitor over a PriorityQueue.
        private PriorityQueue<BcsEvent, int> _queue;
        private readonly object _lock = new object();
        private bool _quit = false;

        public BcsEventLoop() {
            _queue = new PriorityQueue<BcsEvent, int>();
        }

        public void PostEvent(BcsEvent e) {
            lock (_lock) {
                // PriorityQueue dequeues lowest value first, so invert priority
                _queue.Enqueue(e, -e.Priority);
                System.Threading.Monitor.Pulse(_lock);
            }
        }

        public void Exec() {
            lock (_lock) {
                while (!_quit) {
                    while (_queue.Count == 0 && !_quit) {
                        System.Threading.Monitor.Wait(_lock);
                    }

                    if (_quit) break;

                    BcsEvent e = _queue.Dequeue();

                    System.Threading.Monitor.Exit(_lock);
                    // Dispatch event
                    System.Threading.Monitor.Enter(_lock);
                }
            }
        }

        public void Quit() {
            lock (_lock) {
                _quit = true;
                System.Threading.Monitor.PulseAll(_lock);
            }
        }
    }
}

File: ports/go/src/core/kernel/bcs_event.go
package kernel

import (
	"container/heap"
	"sync"
)

// BcsEvent represents the base class for all framework events
type BcsEvent struct {
	Type     int
	Priority int // Higher number means higher priority
}

func NewBcsEvent(eventType int) *BcsEvent {
	return &BcsEvent{Type: eventType, Priority: 0}
}

func NewBcsEventWithPriority(eventType int, priority int) *BcsEvent {
	return &BcsEvent{Type: eventType, Priority: priority}
}

// BcsObject is the base class for all event-receiving objects
type BcsObject struct {
	Parent       *BcsObject
	Children     []*BcsObject
	eventFilters []func(*BcsEvent) bool
}

func NewBcsObject(parent *BcsObject) *BcsObject {
	obj := &BcsObject{Parent: parent, eventFilters: make([]func(*BcsEvent) bool, 0)}
	if parent != nil {
		parent.Children = append(parent.Children, obj)
	}
	return obj
}

func (o *BcsObject) InstallEventFilter(filter func(*BcsEvent) bool) {
	o.eventFilters = append(o.eventFilters, filter)
}

func (o *BcsObject) Event(e *BcsEvent) bool {
	// First process event filters
	for _, filter := range o.eventFilters {
		if filter(e) {
			return true // Handled
		}
	}
	// Base event processing
	return false
}

// priorityQueue implementation for BcsEventLoop using container/heap
type priorityQueue []*BcsEvent

func (pq priorityQueue) Len() int { return len(pq) }
func (pq priorityQueue) Less(i, j int) bool {
	// We want higher priority to be popped first, meaning it needs to be at the front of the heap
	return pq[i].Priority > pq[j].Priority
}
func (pq priorityQueue) Swap(i, j int) { pq[i], pq[j] = pq[j], pq[i] }
func (pq *priorityQueue) Push(x interface{}) {
	*pq = append(*pq, x.(*BcsEvent))
}
func (pq *priorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	*pq = old[0 : n-1]
	return item
}

// BcsEventLoop provides the execution loop for event dispatching
type BcsEventLoop struct {
	queue priorityQueue
	mu    sync.Mutex
	cond  *sync.Cond
	quit  bool
}

func NewBcsEventLoop() *BcsEventLoop {
	loop := &BcsEventLoop{
		queue: make(priorityQueue, 0),
		quit:  false,
	}
	loop.cond = sync.NewCond(&loop.mu)
	return loop
}

func (l *BcsEventLoop) PostEvent(e *BcsEvent) {
	l.mu.Lock()
	defer l.mu.Unlock()

	heap.Push(&l.queue, e)

	l.cond.Signal()
}

func (l *BcsEventLoop) Exec() {
	l.mu.Lock()
	defer l.mu.Unlock()

	for !l.quit {
		for len(l.queue) == 0 && !l.quit {
			l.cond.Wait()
		}

		if l.quit {
			break
		}

		// Pop highest priority event
		e := heap.Pop(&l.queue).(*BcsEvent)

		l.mu.Unlock()
		// Dispatch event to application roots
		_ = e
		l.mu.Lock()
	}
}

func (l *BcsEventLoop) Quit() {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.quit = true
	l.cond.Broadcast()
}
